File: leetcode/Remove_Duplicates_from_Sorted_List_II.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) return null;
        ListNode ret = null;
        ListNode tail = null;
        ListNode prev = head;
        int ct = 1;
        ListNode cur = prev.next;
        while (cur != null) {
            if (cur.val == prev.val) {
                ++ct;
            } else {
                if (ct == 1) {
                    if (ret == null) {
                        ret = tail = prev;
                    } else {
                        tail.next = prev;
                        tail = tail.next;
                    }
                }
                prev = cur;
                ct = 1;
            }
            cur = cur.next;
        }
        if (ct == 1) {
            if (ret == null) {
                ret = tail = prev;
            } else {
                tail.next = prev;
                tail = tail.next;
            }
        }
        if (tail != null) tail.next = null;
        return ret;
    }
}

File: leetcode/Remove_Duplicates_from_Sorted_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode *pre = NULL;
        ListNode *cur = head;
        head = NULL;
        while (cur != NULL) {
            if (cur->next == NULL || cur->val != cur->next->val) {
                // Keep the current node.
                if (head == NULL)
                    head = cur;
                if (pre != NULL)
                    pre->next = cur;
                pre = cur;
                cur = cur->next;
                pre->next = NULL;
            } else {
                // Current value is duplicated. Skip them.
                int dup_val = cur->val;
                while (cur != NULL && cur->val == dup_val) {
                    ListNode *deleting = cur;
                    cur = cur->next;
                    delete deleting;
                }
            }
        }
        return head;
    }
};
